pkg/volume/portworx/portworx_util.go

```diff
@@ -0,0 +1,168 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package portworx
+
+import (
+	"github.com/golang/glog"
+	osdapi "github.com/libopenstorage/openstorage/api"
+	osdclient "github.com/libopenstorage/openstorage/api/client"
+	volumeclient "github.com/libopenstorage/openstorage/api/client/volume"
+	osdspec "github.com/libopenstorage/openstorage/api/spec"
+	osdvolume "github.com/libopenstorage/openstorage/volume"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	osdMgmtPort      = "9001"
+	osdDriverVersion = "v1"
+	pxdDriverName    = "pxd"
+	pwxSockName      = "pwx"
+)
+
+type PortworxVolumeUtil struct {
+	portworxClient *osdclient.Client
+}
+
+// CreateVolume creates a Portworx volume.
+func (util *PortworxVolumeUtil) CreateVolume(p *portworxVolumeProvisioner) (string, int, map[string]string, error) {
+	hostName := p.plugin.host.GetHostName()
+	client, err := util.osdClient(hostName)
+	if err != nil {
+		return "", 0, nil, err
+	}
+
+	capacity := p.options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	// Portworx Volumes are specified in GB
+	requestGB := int(volume.RoundUpSize(capacity.Value(), 1024*1024*1024))
+
+	var labels map[string]string
+	if p.options.CloudTags == nil {
+		labels = make(map[string]string)
+	} else {
+		labels = *p.options.CloudTags
+	}
+
+	specHandler := osdspec.NewSpecHandler()
+	spec, err := specHandler.SpecFromOpts(p.options.Parameters)
```

Comment:
The SpecHandler instance will validate all params and return an error if any. The parameters are validated w.r.t Portworx's Volume Spec.